src/dataflow/src/source/loki.rs

```diff
@@ -21,48 +32,94 @@ struct ConnectionInfo {
+    ) -> Result<reqwest::Response, reqwest::Error> {
         let client = reqwest::Client::new();
-            client.get(format!("{}/loki/api/v1/query_range", conn_info.endpoint))
-                .basic_auth(conn_info.user, Some(conn_info.pw))
-                .query(&[("query", "{job=\"systemd-journal\"}")])
-                .query(&[("start", format!("{}", start))])
-                .query(&[("end", format!("{}", end))])
-                .query(&[("direction", "forward")])
-                .send()
-                .await
+        client
+            .get(format!("{}/loki/api/v1/query_range", conn_info.endpoint))
+            .basic_auth(conn_info.user, Some(conn_info.pw))
+            .query(&[("query", query)])
+            .query(&[("start", format!("{}", start))])
+            .query(&[("end", format!("{}", end))])
+            .query(&[("direction", "forward")])
+            .send()
+            .await
     }
 
-    fn new_stream(self) -> impl stream::Stream<Item = Result<QueryResult, reqwest::Error>> {
-        let polls = IntervalStream::new(tokio::time::interval(self.batch_window));
+    fn new_stream(
+        query: String,
+        batch_window: Duration,
+        conn_info: ConnectionInfo,
+    ) -> impl stream::Stream<Item = Result<QueryResult, reqwest::Error>> {
+        let polls = IntervalStream::new(tokio::time::interval(batch_window));
 
-        let batch_window = self.batch_window;
-        let conn_info = self.conn_info.clone();
+        let conn_info = conn_info;
         polls.then(move |_tick| {
             let start = SystemTime::now() - batch_window;
             let end = SystemTime::now();
 
             Self::query(
                 conn_info.clone(),
                 start.duration_since(UNIX_EPOCH).unwrap().as_nanos(),
-                end.duration_since(UNIX_EPOCH).unwrap().as_nanos())
-                .and_then(|resp| { resp.json::<QueryResult>() })
+                end.duration_since(UNIX_EPOCH).unwrap().as_nanos(),
+                query.clone(),
+            )
+            .and_then(|resp| resp.json::<QueryResult>())
         })
     }
 }
 
 #[async_trait]
 impl SimpleSource for LokiSourceReader {
-
     async fn start(mut self, timestamper: &Timestamper) -> Result<(), SourceError> {
-        let stream = self.new_stream();
+        let stream = Self::new_stream(self.query, self.batch_window, self.conn_info);
+        tokio::pin!(stream);
+
+        while let Some(entry) = stream.next().await {
+            match entry {
+                Ok(result) => {
+                    let Data::Streams(streams) = result.data;
+
+                    let lines: Vec<String> = streams
+                        .iter()
+                        .flat_map(|s| {
+                            s.values.iter().map(|v| {
+                                json!({"line": v.line.clone(), "labels": s.labels}).to_string()
+                            })
+                        })
+                        .collect();
```

Comment:
We can avoid some allocations here using a struct:

```suggestion
                    #[derive(Debug, Serialize)]
                    struct LokiRow<'a> {
                        line: &'a str,
                        labels: &'a HashMap<String, String>,
                    }

                    let lines: Vec<String> = streams
                        .iter()
                        .flat_map(|s| {
                            s.values.iter().map(|v| {
                                serde_json::to_string(&LokiRow {
                                    line: &v.line,
                                    labels: &s.labels,
                                })
                                .unwrap()
                            })
                        })
                        .collect();
```